src/main/java/flaxbeard/steamcraft/handler/SteamcraftEventHandler.java

```diff
@@ -1776,6 +2389,154 @@ public void openMerchant(EntityInteractEvent event) {
+     * and returns the "first" result.
+     * @param player The player
+     * @return The EntityLivingBase near the player.
+     */
+    private EntityLivingBase getEntityFromPlayer(EntityPlayer player) {
+        Vec3 vec = player.getLookVec();
+        double x = vec.xCoord + player.posX;
+        double y = vec.yCoord + player.posY;
+        double z = vec.zCoord + player.posZ;
+
+        AxisAlignedBB aabb = AxisAlignedBB.getBoundingBox(x - 5, y - 5, z - 5, x + 5, y + 5, z + 5);
+        List entities = player.worldObj.getEntitiesWithinAABBExcludingEntity(player, aabb);
+        for (Object object : entities) {
+            if (object instanceof EntityLivingBase) {
+                EntityLivingBase target = (EntityLivingBase) object;
+                if (player.canEntityBeSeen(target) && target.canBeCollidedWith()) {
+                    return target;
+                }
+            }
+        }
+        return null;
+    }
+
+    @SubscribeEvent
+    public void updateBlockBreakSpeed(PlayerEvent.BreakSpeed event) {
+        ItemStack equipped = event.entityPlayer.getCurrentEquippedItem();
+        Block block = event.block;
+        EntityPlayer player = event.entityPlayer;
+        if (equipped != null && equipped.getItem() != null && block != null) {
+            float hardness = block.getBlockHardness(player.worldObj, event.x, event.y, event.z);
+            float newSpeed = 0.0F;
+            if (equipped.getItem() instanceof ItemSteamDrill) {
+                ItemSteamDrill drill = (ItemSteamDrill) equipped.getItem();
+                if (drill.isWound(equipped)) {
+                    if (drill.hasUpgrade(equipped, SteamcraftItems.bigDrill)) {
+                        newSpeed = event.originalSpeed * ((hardness * 1.5F) / 8);
+                    }
+                    if (drill.hasUpgrade(equipped, SteamcraftItems.internalProcessingUnit)) {
+                        if (newSpeed == 0.0F) {
+                            newSpeed = event.originalSpeed / 2;
+                        } else {
+                            newSpeed /= 2;
+                        }
+                    }
+                    if (drill.hasUpgrade(equipped, SteamcraftItems.thermalDrill) ||
+                      drill.hasUpgrade(equipped, SteamcraftItems.chargePlacer)) {
+                        if (newSpeed == 0.0F) {
+                            newSpeed = event.originalSpeed * 5;
+                        } else {
+                            newSpeed *= 5;
+                        }
+                    }
+                }
+            } else if (equipped.getItem() instanceof ItemSteamAxe) {
+                ItemSteamAxe axe = (ItemSteamAxe) equipped.getItem();
+                if (axe.isWound(equipped) && axe.hasUpgrade(equipped, SteamcraftItems.leafBlower)) {
+                    newSpeed = event.originalSpeed / 5F;
+                }
+                if (axe.isWound(equipped) && axe.hasUpgrade(equipped, SteamcraftItems.treeFeller)) {
```

Comment:
It appears I forgot to decrease the mining speed when the Chainsaw is installed.